File: p11/p1143/LongestCommonSubsequence.java
package leetcode.p11.p1143;

import java.util.Arrays;

public class LongestCommonSubsequence {
    /**
     * DP
     * Time complexity: O(mn)
     * Space complexity: O(mn) (Space optimized: O(min(m,n)))
     */
    public int longestCommonSubsequence1(String text1, String text2) {
        int[][] dp = new int[text1.length() + 1][text2.length() + 1];
        for (int i = 0; i < dp.length; i++) {
            dp[i][0] = 0;
        }
        Arrays.fill(dp[0], 0);
        for (int i = 1; i < dp.length; i++) {
            for (int j = 1; j < dp[0].length; j++) {
                if (text1.charAt(i - 1) == text2.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                } else {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }
        return dp[text1.length()][text2.length()];
    }
}

File: p11/p1143/longest_common_subsequence.py
class Solution:
    """
    Feb 15, 2025 17:44
    Time Complexity: O(mn)
    Space Complexity: O(mn)
    """
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        dp = []
        m = len(text1) + 1
        n = len(text2) + 1
        for i in range(m):
            dp.append([0] * n)

        for i in range(1, m):
            for j in range(1, n):
                if text1[i - 1] == text2[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1] + 1
                else:
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

        return dp[-1][-1]
